odoo/cli/upgrade_code.py

```diff
@@ -1,40 +1,80 @@
 #!/usr/bin/env python3
+"""
+Mass-rewrite the entire source code using the scripts found at
+/odoo/upgrade_code
+
+Each script is named {version}-{name}.py and exposes an upgrade function
+that takes a single argument, the file_manager, and returns nothing.
+
+The file_manager acts as a list of files, files have 3 attributes:
+* path: the pathlib.Path where the file is on the file system;
+* addon: the odoo addon in which the file is;
+* content: the re-writtable content of the file (lazy).
+
+There are additional utilities on the file_manager, such as:
+* print_progress(current, total)
+
+Exemple:
+
+    def upgrade(file_manager):
+        files = [f for f in file_manager if f.path.suffix == '.py']
+        for fileno, file in enumerate(files):
+            file.content = file.content.replace(..., ...)
+            file_manager.print_progress(fileno, len(files))
```

Comment:
Couldn't we get rid of `print_progress` altogether by making the iterator of `file_manager` print the progress?  Maybe we could provide a specific iteration method on `file_manager` for that purpose...  Note that the progress is computed with respect to all the files matched by the glob, so it behaves slightly differently.  But that's a good enough approximation IMO.
```suggestion
    def upgrade(file_manager):
        for file in file_manager.iter_files():  # this iterator manages the progress
            if file.path.suffix == '.py':
                file.content = file.content.replace(..., ...)
```
I am not sure it is a good idea, but I thought it was worth sharing :wink:

Reply:
By the way, the special iterator method could even take a parameter to filter files, and the progress could be computed with respect to the filtered files. :thinking:

Reply:
I've a feeling explicit > implicit, it is a single line of code, and it is not too bad if you forget it